accepted/rfc0004-logs.md

```diff
@@ -0,0 +1,203 @@
+
+## Implementing `Storage`
+
+The `Storage` structure is a linked list of vectors (capacitated).
+The idea is that most of the time you just push the event with a very little cost.
+However should you log a lot of events you will have another block re-allocated for you, still in O(1) complexity (well, plus the cost of
+the memory allocation).
+
+It contains an `UnsafeCell` since threads write events to it but someone is going to read the logs and save them.
+Since the vectors are never re-allocated reading a block is always safe. We could have problems
+with lists pointers but I use an atomic counter for the number of blocks which is incremented only after a pointer is changed.
+So, this should hopefully also be safe.
+
+Currently logs are never erased. Calling `save_logs` several times will re-save some tasks several times.
+
+## Implementing `join`
+
+We changed the join function in the following way:
+
+- we use the `in_worker` function to get access to the current thread
+- we go up to the registry and use the thread index to push events in the corresponding storage
+- `op_a` and `op_b` are hijacked to log a new task start at their start and a new task end at their end
+- the current task is ended just before the join and restarted just after
+- we tag the links in the graph between the tasks
+
+# Unresolved questions
+
+## Crates Dependencies
+
+Currently I depend on the following crates. Should I remove these dependencies ?
+- time (easy to remove)
+- itertools (harder to remove but ok)
+- serde (also easy to remove since I output logs in json)
+
+## Svg
+
+All graph algorithms and svg display are not included.
+Is it a good idea ? This means you cannot view the svg display for logs without
+installing `rayon-logs`.
+
+Alternatively I can also remove the `RunLog` type and file and save logs in an even cruder format
+to keep more processing in `rayon-logs`. If we keep the crates separated this might actually be a good idea.
+
+## Logged Iterator
+
+There is a convenient `LoggingIterator` in `rayon-logs` for calling `subgraph` automatically (with automatic size)
+on each terminal task of a `ParallelIterator`. I could but it here but it is not completely sound (should'nt be zipped for example).
+Well, since you can also easily map to call `subgraph` it might not be a necessity.
+
+## Sharing structs with `rayon-logs`
+
+What is the way to go for sharing structs with `rayon-logs` (avoid the type re-definition) ?
+I would rather have the events types public in rayon but it could be the other way round.
+Note that this problem disappears if the svg code is moved inside rayon.
+
+## Optimizing raw events types.
+
+We could maybe rewrite the raw events to decrease their sizes (and do more push).
+I'm not sure it is a big deal though since all costs are dominated by the thread local storage.
+
+## Saving logs.
+
+It is the user's responsibility to not save logs while using threads. It should not crash but the graphs obtained

```

Comment:
How are logs saved? I guess there is a function that gets called by the user to save logs?

Reply:
yes, there is a save_logs function (global pool) or method on the pool.